### PR TITLE
rm: Add documentation for event queues

### PR DESCRIPTION
**Pull request by doanac**
Signed-off-by: Andy Doan <andy@foundries.io>

### CODE REVIEW

**Comment by doanac** [comment]
there will be a broken link until i merge some corresponding fioctl changes. I wanted to get people looking through the docs first in case we found a need to re-do the fioctl commands.

**Comment by doanac** [comment]
Docs for b57a37c146409f463d120f5eda362cdf2549ab81 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1400/docs/artifacts/html/index.html

File: source/reference-manual/factory/event-queues.rst [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+DEVICE_OTA_COMPLETED
+~~~~~~~~~~~~~~~~~~~~
+::
+
+ {
+   "Uuid": <string: DEVICE_UUID>,
+   "Time": <integer: unix seconds>,
+   "Target": <string: target name>,
+   "Id": <string: update correlation-id>,
+   "Success": <boolean>
```
Comment (mike-sul) [inline source/reference-manual/factory/event-queues.rst:157]: Maybe we should add a reason of an update failure as well as a cause that trigerred an update (e.g. sync, force update, just regular update, etc)

**Review by mike-sul — COMMENTED** [review]
Is the ota-lite and fioctl parts are deployed? It would be great if we try  the commands stated in the doc.

Also, the auth part for the push queue is really confusing to me. I re-read this doc https://cloud.google.com/pubsub/docs/push#authentication twice and still don't get it :).

**Comment by doanac** [comment]
> Is the ota-lite and fioctl parts are deployed? It would be great if we try the commands stated in the doc.

ota-lite is in my staging and the fioctl branch is up. I was holding off on completing those until someone had a look at this.

**Comment by doanac** [comment]
Docs for 6e661cac932c382a776b86b8ed3f95c0ebdc0657 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1402/docs/artifacts/html/index.html

File: source/reference-manual/factory/event-queues.rst [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+     "publishTime":"2022-03-30T15:18:21.095Z",
+     "publish_time":"2022-03-30T15:18:21.095Z"
+   },
+   "subscription":"projects/osf-prod/subscriptions/xxxxxxxx"
+  }
+
+Push Queue Security
+~~~~~~~~~~~~~~~~~~~
+Incoming requests will include a header, ``Authorization: Bearer <jwt>``.
+This JWT is signed by one of Google's published `public keys`_.
```
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst:110]: I should say signed by one of "Google's published keys". It's asymmetric encryption and google publishes their public keys.

There is no user provided key. Google does the encryption.

File: source/reference-manual/factory/event-queues.rst [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+The JWT audience header is set to the URL you specified when creating
```
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst:111]: I include an example source file that does that

File: source/reference-manual/factory/event-queues.rst [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+this API is fully functional, it can be hard for certain operators to
+build their custom tooling for large numbers of devices if they
+must constantly poll the API for information. Event queues were
+designed to solve this issue.
+
+Event queues provide a way for customers to receive messages about
+important events that aren't easy to poll for. These include things
+like:
+
+ * The first time a device connects to the device gateway.
```
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst]: 9136821d8bc19175dc11ca56e0ee0ed9bcf47b0b

File: source/reference-manual/factory/event-queues.rst [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+ * When a device applies a configuration change.
+ * When an OTA starts and completes.
+
+Event queues are implemented on Google's PubSub_ to provide customers
```
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst]: 64ec84ad580ac44f1123ed5bebd4ca07ad870565

File: source/reference-manual/factory/event-queues.rst [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+with a well understood and tested framework. There are two types
+of event queues that can be created:
+
+ * Push - Works like a traditional webhook service. Events are sent
```
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst]: 64ec84ad580ac44f1123ed5bebd4ca07ad870565

**Comment by doanac** [comment]
Docs for 445e50b2466d43bdbe1abe1d789cc260fdc9498e are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1403/docs/artifacts/html/index.html

File: source/reference-manual/factory/event-queues.rst [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+   to a URL managed by the customer where they can ingest events.
+
+ * Pull - Works like a tradition message queue system where the
```
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst]: 445e50b2466d43bdbe1abe1d789cc260fdc9498e

File: source/reference-manual/factory/event-queues.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst]: b7730120e52f434295b42d3e34de3bda41d840bb

**Comment by doanac** [comment]
Docs for b7730120e52f434295b42d3e34de3bda41d840bb are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1404/docs/artifacts/html/index.html

**Comment by mike-sul** [comment]
Maybe we should add some info on a retention policy, could be just one sentence and reference to the corresponding PubSub doc.

**Comment by mike-sul** [comment]
Not related to the doc. The push security seems a bit odd to me. I would expect that each user will have their own key pairs.

**Comment by doanac** [comment]
> Maybe we should add some info on a retention policy, could be just one sentence and reference to the corresponding PubSub doc.

fee70f05dd252b84a6dc5c59feb183ac011a7aa0

**Comment by doanac** [comment]
Docs for fee70f05dd252b84a6dc5c59feb183ac011a7aa0 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1405/docs/artifacts/html/index.html

File: source/reference-manual/factory/event-queues.rst [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+Event queues provide a way for customers to receive messages about
+important events that aren't easy to poll for. These include things
+like:
+
+ * The first time a device connects to the :ref:`device gateway <ref-ota-architecture>`.
+ * When a device applies a configuration change.
+ * When an OTA update starts and completes.
+
+Event queues are implemented using Google PubSub_ to provide customers
+with a well understood and tested framework. There are two types
```
Comment (kprosise) [inline source/reference-manual/factory/event-queues.rst:22]:
```suggestion
Event queues are implemented using Google PubSub_ to provide
a well understood and tested framework. There are two types
```
No need to specify customers given the context.

File: source/reference-manual/factory/event-queues.rst [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+of event queues that can be created:
+
+ * Push - Works works as a webhook_ service. Events are sent
+   to a URL managed by the customer where they can process events.
```
Comment (kprosise) [inline source/reference-manual/factory/event-queues.rst:26]:
```suggestion
   to a managed URL where they can be processed.
```
Events already referenced within sentence

File: source/reference-manual/factory/event-queues.rst [hunk 10]
```diff
@@ -18,12 +18,12 @@ like:
...
  * When an OTA update starts and completes.
 
-Event queues are implemented on Google's PubSub_ to provide customers
+Event queues are implemented using Google PubSub_ to provide customers
 with a well understood and tested framework. There are two types
 of event queues that can be created:
 
- * Push - Works like a traditional webhook service. Events are sent
-   to a URL managed by the customer where they can ingest events.
+ * Push - Works works as a webhook_ service. Events are sent
```
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst:25]: 66bfd8e4c020fa3021b4f800dc9d8342ffc81a6a

File: source/reference-manual/factory/event-queues.rst [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+
+Event queues are implemented using Google PubSub_ to provide customers
+with a well understood and tested framework. There are two types
+of event queues that can be created:
+
+ * Push - Works works as a webhook_ service. Events are sent
+   to a URL managed by the customer where they can process events.
+
+ * Pull - Works like a typical message queue system where the
+   customer can write their own client to receive and process events.
```
Comment (kprosise) [inline source/reference-manual/factory/event-queues.rst:29]:
```suggestion
 * Pull - Works like a typical message queue system where one
   can write their own client to receive and process events.
```

File: source/reference-manual/factory/event-queues.rst [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+
+PubSub documentation includes a very useful guide_ to help decide
+which approach will work best for you. They also include a wide
+range of `client libraries`_ for consuming the Pull API. PubSub
+Subscriptions are created with default retention, expiration,
```
Comment (kprosise) [inline source/reference-manual/factory/event-queues.rst:34]:
```suggestion
subscriptions are created with default retention, expiration,
```

**Review by MrCry0 — APPROVED** [review]

File: source/reference-manual/factory/event-queues.rst [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+and acknowledgement values_.
+
+Implementation Details
+----------------------
+
+Each FoundriesFactory is given a single PubSub Topic. Each Push
+and Pull queue created by a customer results in the creation of
+a PubSub Subscription. The FoundriesFactory API just provides a
```
Comment (kprosise) [inline source/reference-manual/factory/event-queues.rst:42]:
```suggestion
a PubSub Subscription. The FoundriesFactory API provides a
```
Just can be excluded without lose of meaning.

File: source/reference-manual/factory/event-queues.rst [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+---------------------
+
+A push queue requires a little up front work:
+
+ * Having an internet facing server to run the webhook service.
+ * Authenticating incoming requests.
+
+Quick Start Example
+~~~~~~~~~~~~~~~~~~~
+A great way to prototype a push queue by using ngrok_. This tool allows
```
Comment (kprosise) [inline source/reference-manual/factory/event-queues.rst:69]:
```suggestion
A great way to prototype a push queue is by using ngrok_. This tool allows
```

**Review by kprosise — APPROVED** [review]
Left some suggestions.

**Comment by doanac** [comment]
Docs for 151e9676fd220de8dde2d626c157bca7a87312ab are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1409/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 00ed5f10b9c20a0d7b7eeb0539182432644deb91 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1410/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for a1c727fb1e148d885b82d225c4714ab92978436d are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1411/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for d180c65c508f00fc484e58f46d13ca2fb8bc1d5b are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1412/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 479abedc17e92995ce7fa1955e4bd5132d56ef0f are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1414/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 450edc7d3f4b33c8a0ddb411780a0b95b976fc29 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1413/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for d592ac1a96278fac7c60fb552c413c5ad1ba0f10 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1415/docs/artifacts/html/index.html

**Comment by doanac** [comment]
Docs for 00dc650164c580b766484e79801ce6bf86fd8fae are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1416/docs/artifacts/html/index.html

File: source/reference-manual/factory/event-queues.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply source/reference-manual/factory/event-queues.rst]: This auth schema guarantees that a request is coming from GC PubSub. But when GC PubSub is used in multi-tenant manner how to guarantee that a request (webhook) intended for one customer is not sent to and authorized by another customer.

Anyway, it's not related to the doc itself.

**Review by mike-sul — APPROVED** [review]
LGTM, eager to try the instruction provided in the doc.

File: source/reference-manual/factory/event-queues.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (doanac) [reply source/reference-manual/factory/event-queues.rst]: > how to guarantee that a request (webhook) intended for one customer is not sent to and authorized by another customer.

Each tenant has their own queue. We create IAM rules that isolate their service accounts to only have access to topic. Then our publishing code (like the rest of our solution) has to handle multiple tenants correctly.

[the remark this answers is not in the document]
Comment (mike-sul) [reply source/reference-manual/factory/event-queues.rst]: I understand that. But mechanism itself is not protected from incorrect implementation or malicious user on our/publishing side :).

**Comment by doanac** [comment]
@kprosise - added a small note about event delays.

**Comment by doanac** [comment]
Docs for 72bef2c89475293162b3e95391036aaaf3d4706f are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1417/docs/artifacts/html/index.html

**Comment by doanac** [comment]
ci-retest

**Comment by doanac** [comment]
@kprosise - once CI passes, this change should be good to merge. we are deploying all the changes to production today

**Comment by doanac** [comment]
Docs for 72bef2c89475293162b3e95391036aaaf3d4706f are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1421/docs/artifacts/html/index.html

**Comment by doanac** [comment]
@kprosise - this link-check error:
```
Warning, treated as error:
/repo/source/reference-manual/factory/event-queues.rst.rst:72:broken link: ../../_static/push_queue_example.py ()
```
Should i just put that in the `conf.py:linkcheck_ignore` or is there a better way to handle that?

**Comment by kprosise** [comment]
> @kprosise - this link-check error:
> 
> ```
> Warning, treated as error:
> /repo/source/reference-manual/factory/event-queues.rst.rst:72:broken link: ../../_static/push_queue_example.py ()
> ```
> 
> Should i just put that in the `conf.py:linkcheck_ignore` or is there a better way to handle that?

@doanac I want to say there is a better way, let me take a closer look and I'll get back to you.

**Comment by kprosise** [comment]
@doanac As I thought, we could instead use the download role, changing the two links to the file to ``:download:`example push queue<../../_static/push_queue_example.py>` ``. On the other hand, this puts a little download file icon next to it, and is not as clean looking in the rst. Either solution works.

**Comment by doanac** [comment]
@kprosise - i like your idea. It's also nice to indicate to the user its a download ahead of time. 

https://github.com/foundriesio/docs/pull/357/commits/a3120e87f73e20b587e92ac6618c9c18d24b6552

**Comment by doanac** [comment]
Docs for a3120e87f73e20b587e92ac6618c9c18d24b6552 are browsable at: https://ci.foundries.io/projects/fio-docs/builds/1423/docs/artifacts/html/index.html

**Comment by kprosise** [comment]
@doanac Care if I go ahead and merge it?

**Comment by doanac** [comment]
@kprosise - please do

**Pull request merged**